platform-sdk/swirlds-jasperdb/src/main/java/com/swirlds/merkledb/files/DataFileWriter.java

```diff
@@ -163,64 +229,106 @@ public synchronized long writeCopiedDataItem(final long serializedVersion, final
     /**
      * Store data item in file returning location it was stored at.
      *
-     * @param dataItem
-     * 		the data item to write
+     * @param dataItem the data item to write
      * @return the data location of written data in bytes
-     * @throws IOException
-     * 		if there was a problem appending data to file
+     * @throws IOException if there was a problem appending data to file
      */
     public synchronized long storeDataItem(final D dataItem) throws IOException {
-        // find offset for the start of this new data item, we assume we always write data in a whole number of blocks
-        final long byteOffset = writePosition;
+        // find offset for the start of this new data item, we assume we always write data in a
+        // whole number of blocks
+        final int currentWritingMmapPos = writingMmap.position();
+        final long byteOffset = mmapPositionInFile + currentWritingMmapPos;
         // write serialized data
-        final int totalDataWritten = dataItemSerializer.serialize(dataItem, writingStream);
-        // update write position
-        writePosition += totalDataWritten;
+        writeWithRetry(() -> dataItemSerializer.serialize(dataItem, writingMmap), null);
         // increment data item counter
         dataItemCount++;
         // return the offset where we wrote the data
         return DataFileCommon.dataLocation(index, byteOffset);
     }
 
+    /** A helper method to write a byte buffer to the file. */
+    private void writeBytes(final ByteBuffer data) throws IOException {
+        final int pos = data.position();
+        final int limit = data.limit();
+        writeWithRetry(() -> writingMmap.put(data), () -> {
```

Comment:
We could elide the creation of at least one of the runnables by using method handles here, but not sure it's enough of an impact to worry about?

Reply:
It'd love to get rid of RunnableWithIOException and use something standard like Consumer, but it wouldn't work for the remaining two methods. Use two different code paths for writeBytes() and store/copyDataItem() doesn't look good either.

Reply:
I had to revert the changes with RunnableWithIOException and writeWithRetry() after I got benchmark results from Oleg. He observed redundant memory allocations, especially in writeCopiedDataItem().